torch/csrc/jit/dynamic_dag.h

```diff
@@ -0,0 +1,561 @@
+#pragma once
+
+#include <algorithm>
+#include <iostream>
+#include <type_traits>
+#include <unordered_set>
+#include <vector>
+
+#include "torch/csrc/utils/functional.h"
+
+namespace torch { namespace jit { namespace detail {
+
+// DynamicDAG is a simple directed acyclic graph that dynamically maintains a
+// topological order as edges/vertices are added and removed.
+//
+// [Example applications]
+// - Let's say you have a DAG where each vertex is black or red. How do we
+//   merge black nodes that are directly connected by contracting the
+//   edge between them while still maintaining the DAG and a topological order?
+//   Use contractEdge().
+// - Let's say you have a DAG where each vertex is a Node* and the edges represent
+//   data dependencies. We wish to determine if adding a new Node* with certain
+//   data dependencies (or moving an existing one to use new dependencies) is valid.
+//   Use DynamicDAG::addEdge() to add the new data dependencies to the DAG:
+//   it will either find a valid reordering of the DAG's topological order or throw
+//   if the resulting DAG is invalid.
+//
+// The implementation is based off of the PK algorithm in the following paper:
+// "A Dynamic Topsort Algorithm for Directed Acyclic Graphs"
+// by David Pearce and Paul Kelly
+// https://www.doc.ic.ac.uk/~phjk/Publications/DynamicTopoSortAlg-JEA-07.pdf
+// It is summarized in [Edge addition] (see DynamicDAG<T>::addEdge)
+
+template <typename T> struct Vertex;
+template <typename T> struct DynamicDAG;
+template <typename T> using vertex_list = std::vector<Vertex<T>*>;
+template <typename T> using unique_vertex = std::unique_ptr<Vertex<T>>;
+
+enum DFSDirection {forward, backward};
+
+// Used to represent adjacency lists in DynamicDAG.
+// NOT a multiset.
+// Because our graphs shouldn't fan out or in very much,
+// we use std::vector<Vertex<T>*> to record edges.
+// In all of the complexity analysis it is assumed that
+// inserting, erasing, and finding take constant time.
+template <typename T>
+struct ordered_vertex_set {
```

Comment:
What's ordered about it? I thought we don't give any guarantees.

Reply:
I'd really much rather call it a `multiset`

Reply:
It's ordered in the sense that there is some ordering to its elements (it matters what order you iterate through the elements).

ordered_vertex_set isn't used as a multiset -- I can add asserts to insert, erase, and contains so that this data structure doesn't get used as a multiset

Reply:
Not being used as a multiset doesn't mean that this data structure doesn't have multiset semantics (this is very important!!).

I wouldn't rely on the ordering all that much since it can be arbitrary unless you call `sort()` before you iterate over it.

Reply:
Ah I see. Right now, this data structure doesn't completely have multiset semantics: `erase()` is implemented incorrectly for a multiset because it just erases one element and not all elements, so I am wary of naming it something that it is not.

I'd like this data structure to not have multi-set semantics so I think I will add checks on the insert for uniqueness.